src/common/mfu_io.c

```diff
@@ -9,13 +9,203 @@
+        }
+
+        hdl = calloc(1, sizeof(struct mfu_dir_hdl));
+        if (hdl == NULL)
+		return NULL;
+
+        strncpy(hdl->name, name, PATH_MAX-1);
+        hdl->name[PATH_MAX-1] = '\0';
+
+        rc = dfs_lookup(mfu_file->dfs, name, O_RDWR, &hdl->oh, NULL, NULL);
+	if (rc) {
+		fprintf(stderr, "dfs_lookup() of %s Failed", name);
+		return NULL;
+	}
+
+        rc = d_hash_rec_insert(dir_hash, hdl->name, strlen(hdl->name),
+                               &hdl->entry, true);
+	if (rc)
+		return NULL;
+
+        return hdl->oh;
+}
+
+static int parse_filename(const char* path, char** _obj_name, char** _cont_name)
+{
+	char *f1 = NULL;
+	char *f2 = NULL;
+	char *fname = NULL;
+	char *cont_name = NULL;
+	int rc = 0;
+
+	if (path == NULL || _obj_name == NULL || _cont_name == NULL)
+		return -EINVAL;
+
+	if (strcmp(path, "/") == 0) {
+		*_cont_name = strdup("/");
+		if (*_cont_name == NULL)
+			return -ENOMEM;
+		*_obj_name = NULL;
+		return 0;
+	}
+
+	f1 = strdup(path);
+	if (f1 == NULL) {
+                rc = -ENOMEM;
+                goto out;
+        }
+
+	f2 = strdup(path);
+	if (f2 == NULL) {
+                rc = -ENOMEM;
+                goto out;
+        }
+
+	fname = basename(f1);
+	cont_name = dirname(f2);
+
+	if (cont_name[0] == '.' || cont_name[0] != '/') {
+		char cwd[1024];
+
+		if (getcwd(cwd, 1024) == NULL) {
+                        rc = -ENOMEM;
+                        goto out;
+                }
+
+		if (strcmp(cont_name, ".") == 0) {
+			cont_name = strdup(cwd);
+			if (cont_name == NULL) {
+                                rc = -ENOMEM;
+                                goto out;
+                        }
+		} else {
+			char *new_dir = calloc(strlen(cwd) + strlen(cont_name)
+					       + 1, sizeof(char));
+			if (new_dir == NULL) {
+                                rc = -ENOMEM;
+                                goto out;
+                        }
+
+			strcpy(new_dir, cwd);
+			if (cont_name[0] == '.') {
+				strcat(new_dir, &cont_name[1]);
+			} else {
+				strcat(new_dir, "/");
+				strcat(new_dir, cont_name);
+			}
+			cont_name = new_dir;
+		}
+		*_cont_name = cont_name;
+	} else {
+		*_cont_name = strdup(cont_name);
+		if (*_cont_name == NULL) {
+                        rc = -ENOMEM;
+                        goto out;
+                }
+	}
+
+	*_obj_name = strdup(fname);
+	if (*_obj_name == NULL) {
+		free(*_cont_name);
+		*_cont_name = NULL;
+                rc = -ENOMEM;
+                goto out;
+	}
+
+out:
+	if (f1)
+		free(f1);
+	if (f2)
+		free(f2);
+	return rc;
+}
+#endif
+
 /* calls access, and retries a few times if we get EIO or EINTR */
+int mfu_file_access(const char* path, int amode, mfu_file_t* mfu_file)
+{
+    int rc;
+    if (mfu_file->type == POSIX) {
+        rc = mfu_access(path, amode);
+    } else if (mfu_file->type == DAOS) {
+        rc = daos_access(path, amode);
+    }
+    if (rc != 0) {
+        rc = -1;
+    }
+    return rc;
+}
+
 int mfu_access(const char* path, int amode)
```

Comment:
```suggestion
int mfu_file_access(const char* path, int amode, mfu_file_t* mfu_file)
{
    if (mfu_file->type == POSIX) {
        int rc = mfu_access(path, amode);
        return rc;
    } else if (mfu_file->type == DAOS) {
        int rc = daos_access(path, amode);
        if (rc != 0) {
            rc = -1;
        }
        return rc;
    } else {
        // let's catch this case and abort with an error
    }
}

/* calls access, and retries a few times if we get EIO or EINTR */
int mfu_access(const char* path, int amode)
```

Reply:
Let's change the other mfu_file wrappers to look like this, where we add a catch all case and abort if they mfu_file->type is not known.  Also, let's just return the value from each function and then move the return code and errno to within the daos_access() code rather than straddle it here and there.